### PR TITLE
convert evalu8.c to evalu8.d

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Comment by rainers** [comment]
Building on Appveyor fails because the D part assumes `real` exists on the C++ side, but that uses longdouble_soft.
As this is only a temporary problem until the other C files are translated aswell, it's probably easiest to just force the correct mangling with `pragma(mangle)` on a couple of functions.

**Comment by rainers** [comment]
Affected functions seem to be `el_toldouble`, `_modulo` and `Complex_ld::abs`.

**Comment by rainers** [comment]
Another option is to use `longdouble_soft` in D instead of `real` whenever version CRuntime_Microsoft is active. That needs a tiny change for y2lx, too.

**Comment by WalterBright** [comment]
Since it'll all be converted to D anyway, I figured might as well go ahead and convert bcomplex.c to D in this PR.

**Comment by WalterBright** [comment]
Appveyor failures:
```
unittest
core.exception.AssertError@src\core\math.d(116): unittest failure
----------------
0x01114497
0x011AA5D3 in gc_clrProxy
core.exception.AssertError@src\rt\util\typeinfo.d(250): Assertion failure
----------------
0x01114441
0x011AA5D3 in gc_clrProxy
2/42 unittests FAILED
--- errorlevel 1
```
which have nothing to do with this PR. @rainers what's going on?

**Comment by WalterBright** [comment]
well, maybe they do :-(

File: src/dmd/backend/bcomplex.d [hunk 1]
```diff
@@ -24,28 +22,291 @@ struct Complex_f
...
+        }
+        return c;
+    }
 }
 
+
 struct Complex_ld
 {
-    targ_ldouble re, im;
+    real re, im;
```
Comment (rainers) [inline src/dmd/backend/bcomplex.d]: You have to keep `targ_ldouble` here because when built with LDC on Windows, `real` is the same as `double`.

**Comment by WalterBright** [comment]
@rainers are you able to replicate the AppVeyor failures? I can't on my machine.

File: src/dmd/backend/bcomplex.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/backend/bcomplex.d]: ok

**Comment by WalterBright** [comment]
> You have to keep targ_ldouble here because when built with LDC on Windows, real is the same as double.

Making that change causes semaphoreci to fail with:

```
gdmd -c -of../generated/linux/release/64/divcoeff.o  -version=MARS -fPIC -w -de -O -release -inline -m64  dmd/backend/divcoeff.d
dmd/backend/bcomplex.d:217:1: error: struct dmd.backend.bcomplex.Complex_ld no size yet for forward reference
 struct Complex_ld
 ^
dmd/backend/cdef.d:798:1: error: union dmd.backend.cdef.eve no size yet for forward reference
 union eve
 ^
````

File: src/dmd/e2ir.d [hunk 2]
```diff
@@ -1479,8 +1479,8 @@ elem *toElem(Expression e, IRState *irs)
                     break;
 
                 case TYcldouble:
-                    e.EV.Vcldouble.re = re;
-                    e.EV.Vcldouble.im = im;
+                    e.EV.Vcldouble.re = cast(real)re;
+                    e.EV.Vcldouble.im = cast(real)im;
```
Comment (rainers) [inline src/dmd/e2ir.d]: This cast is no longer necessary and is harmful to the LDC build, too.

Comment (WalterBright) [reply src/dmd/e2ir.d]: ok, gone

File: src/dmd/backend/evalu8.d [hunk 3]
```diff
@@ -1876,4 +1898,51 @@ elem * evalu8(elem *e, goal_t goal)
   return e;
 }
 
-#endif /* !SPP */
+/******************************
+ * This is the same as the one in el.c, but uses native D reals
+ * instead of the soft long double ones.
+ */
+
+extern (D) real el_toldoubled(elem *e)
```
Comment (rainers) [inline src/dmd/backend/evalu8.d]: Sorry, still more uses of `real` instead of `targ_ldouble`...

Comment (WalterBright) [reply src/dmd/backend/evalu8.d]: fixed

File: src/dmd/backend/evalu8.d [hunk 4]
```diff
@@ -1876,4 +1898,51 @@ elem * evalu8(elem *e, goal_t goal)
...
+    }
+    return result;
+}
+
+/***************************************
+ * Copy of _modulo from fp.c. Here to help with linking problems.
+ */
+version (CRuntime_Microsoft)
+{
+    extern (D) private real _modulo(real x, real y)
```
Comment (rainers) [inline src/dmd/backend/evalu8.d]: I think the casts should go into this function instead of the call.

File: src/dmd/backend/evalu8.d [hunk 5]
```diff
@@ -6,46 +6,51 @@
...
-#if _MSC_VER
-#define isnan _isnan
-#endif
-
-#include        "cc.h"
-#include        "oper.h"                /* OPxxxx definitions           */
-#include        "global.h"
-#include        "el.h"
-#include        "type.h"
+version (SPP)
```
Comment (rainers) [inline src/dmd/backend/evalu8.d]: BTW: no module name here? (I won't object to module names without package if that's that's how you want to do it ;-) )

File: src/dmd/backend/evalu8.d (hunk 4 above)
Comment (WalterBright) [reply src/dmd/backend/evalu8.d]: all right

File: src/dmd/backend/evalu8.d (hunk 5 above)
Comment (WalterBright) [reply src/dmd/backend/evalu8.d]: No imports of it yet, I'll probably worry about that when imports happen. Right now, I just want to get it to work :-)

**Comment by rainers** [comment]
Still too many uses of `real`. Shall I push a fix?

**Comment by WalterBright** [comment]
Yes, I could use your help. I'm mostly just floundering about on these platforms I have little knowledge of.

**Comment by rainers** [comment]
> Yes, I could use your help. I'm mostly just floundering about on these platforms I have little knowledge of.

Done. Now let's see the other platforms complain...

**Comment by rainers** [comment]
This failed because comparison of two `targ_ldouble`s is forwarded to `longdouble_soft.opCmp`, but that cannot return "false" for all comparisons involving `nan`. Now added explicit checks.

**Comment by WalterBright** [comment]
It's all green and ready to pull! Thanks @rainers for your invaluable help. ping @JinShil @wilzbach @jacob-carlborg ?

File: src/dmd/backend/bcomplex.d [hunk 6]
```diff
@@ -4,14 +4,15 @@
  *
  * Copyright:   public domain
  * License:     public domain
- * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/backend/bcomplex.d, backend/_bcomplex.d)
+ * Source:      $(DMDSRC backend/_bcomplex.d)
```
Comment (wilzbach) [inline src/dmd/backend/bcomplex.d]: FYI underscore escaping is no longer needed. We enabled the opt-out about two months ago.
See: https://github.com/dlang/dlang.org/pull/2307

File: src/dmd/backend/evalu8.d [hunk 7]
```diff
@@ -82,9 +89,7 @@ int boolres(elem *e)
                 case TYbool:
                 case TYwchar_t:
                 case TYenum:
-#if !MARS
                 case TYmemptr:
-#endif
```
Comment (wilzbach) [inline src/dmd/backend/evalu8.d]: Shouldn't this result in a `version(MARS){} else { ... }` ?

File: src/dmd/backend/evalu8.d [hunk 8]
```diff
@@ -275,33 +282,37 @@ int iffalse(elem *e)
...
     targ_ldouble d1,d2;
-    elem esave;
+    elem esave = void;
+
+    static bool unordered(T)(T d1, T d2) { return isnan(d1) || isnan(d2); }
 
     assert((statusFE() & 0x3800) == 0);
-    assert(e && EOP(e));
-    op = e->Eoper;
+    assert(e && !OTleaf(e.Eoper));
```
Comment (wilzbach) [inline src/dmd/backend/evalu8.d]: Why not define EOP or EBIN as a function?

**Review by JinShil — APPROVED** [review]

File: src/dmd/backend/bcomplex.d (hunk 6 above)
Comment (wilzbach) [reply src/dmd/backend/bcomplex.d]: Also last year you made a point of using absolute links here as they are clickable (see https://github.com/dlang/dmd/pull/7030).
Though I was a fan of DMDSRC, so I don't mind ;-)

**Review by wilzbach — APPROVED** [review]
Nice!

File: src/dmd/backend/evalu8.d (hunk 8 above)
Comment (WalterBright) [reply src/dmd/backend/evalu8.d]: I don't think those macros paid off.

File: src/dmd/backend/evalu8.d (hunk 7 above)
Comment (WalterBright) [reply src/dmd/backend/evalu8.d]: I decided that having too many version statements was clutter, and so removed the ones that weren't really necessary.

File: src/dmd/backend/bcomplex.d (hunk 6 above)
Comment (WalterBright) [reply src/dmd/backend/bcomplex.d]: yes, it should be the full version.

File: src/dmd/backend/bcomplex.d [hunk 9]
```diff
@@ -24,28 +25,291 @@ struct Complex_f
...
+    {
+        Complex_f q;
+        targ_ldouble r;
+        targ_ldouble den;
+
+        if (fabs(y.re) < fabs(y.im))
+        {
+            r = y.re / y.im;
+            den = y.im + r * y.re;
+            q.re = cast(float)((x.re * r + x.im) / den);
```
[the remark this answers is not in the document]
Comment (WalterBright) [reply src/dmd/backend/bcomplex.d]: they make it clearer what is happening

**Comment by JinShil** [comment]
It also appears bcomplex.c was not deleted in this PR.  Shouldn't it be?

File: src/dmd/backend/bcomplex.d (hunk 9 above)
[the remark this answers is not in the document]
Comment (rainers) [reply src/dmd/backend/bcomplex.d]: and the casts are necessary if targ_ldouble resolves to longdouble_soft, as there is no implicit conversion to float/double then.

**Pull request merged**